[unifi] Added feature to be able to block and unblock clients

scurb opened the pull request:
This PR gives the possibility to toggle a switch which sets and unsets the "blocking state" of a client.

In order to give the possibility to also block/unblock wired clients, getClient-function in UniFiControllerThingHandler was changed so it does not return null if wired clients.

Block is added as a R/W channel for the Thing type "unifiwirelessclient".

Link to built JAR: https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.unifi/2.5.0-SNAPSHOT/org.openhab.binding.unifi-2.5.0-SNAPSHOT.jar

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -95,7 +97,7 @@ public void login() throws UniFiException {
         UniFiControllerRequest<UniFiClient[]> req = newRequest(UniFiClient[].class);
         req.setPath("/api/s/" + site.getName() + "/stat/alluser");
-        req.setQueryParameter("within", 24);
+        req.setQueryParameter("within", 168); // scurb: Changed to 7 days.
```
Why did you change it?

scurb replied:
I changed the insight value in order to increase the chances of finding the client that should be blocked/unblocked. If the clients are not active, the binding queries the insight api endpoint (kind of a history table). If a client has been disconnected for more than a day (previous value) then I thought a week should be more fair. I don’t see any negative perspective in increasing the value more then maybe the ”load” on the controller/oh binding in generating or parsing the result.

scurb replied:
Yes, exactly. The client gets disconnected which makes it only visible when doing Insights-lookup. And for clients that has been disconnected I wanted to make it possible to find them even though a couple of days has passed (one week).

scurb commented:
I changed the insight value in order to increase the chances of finding the client that should be blocked/unblocked. If the clients are not active, the binding queries the insight api endpoint (kind of a history table). If a client has been disconnected for more than a day (previous value) then I thought a week should be more fair. I don’t see any negative perspective in increasing the value more then maybe the ”load” on the controller/oh binding in  generating or parsing the result. 

/Patrik

> 10 feb. 2019 kl. 12:13 skrev Martin van Wingerden <notifications@github.com>:
> 
> @martinvw commented on this pull request.
> 
> In addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
> 
> > @@ -95,7 +97,7 @@ public void login() throws UniFiException {
>      public UniFiClient[] getInsights(UniFiSite site) throws UniFiException {
>          UniFiControllerRequest<UniFiClient[]> req = newRequest(UniFiClient[].class);
>          req.setPath("/api/s/" + site.getName() + "/stat/alluser");
> -        req.setQueryParameter("within", 24);
> +        req.setQueryParameter("within", 168); // scurb: Changed to 7 days.
> Why did you change it?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

martinvw commented:
@mgbowman can you take a look / review?

@kaikreuzer can you add @mgbowman to the team?

Hilbrand commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -146,4 +147,11 @@
+		<item-type>Switch</item-type>
+		<label>Blocked</label>
+		<description>Is device blocked</description>
+		<state readOnly="false"></state>
```
State readOnly is default `false`. Therefor remove it here.

scurb replied:
Removed.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiClientThingConfig.java:
```diff
@@ -41,6 +43,10 @@ public int getConsiderHome() {
         return considerHome;
     }
 
+    public boolean getBlocked() {
```
Boolean field method should start with `is`:
```suggestion
    public boolean isBlocked() {
```

scurb replied:
Renamed.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -111,15 +114,34 @@ public void initialize() {
     @Override
     public void handleCommand(ChannelUID channelUID, Command command) {
-        // mgb: only handle the command if we're ONLINE
-        if (getThing().getStatus() == ONLINE) {
```
Why it the channel block called when the device is offline? If the device is off line it can't accept commands I would assume.

scurb replied:
If the client gets blocked it will be put "OFFLINE". In order to get it "unblocked" in has to be able to handle commands to Things that are offline. Later on in the binding (when API is called) it gets data from "Insights" which is the "offline"/history database of disconnected/offline clients.

Hilbrand replied:
Are you sure about this? If I block a device it stays online, but the channel `CHANNEL_ONLINE` becomes off. In other words. Don't you mix up between thing status online/offline and the channel `CHANNEL_ONLINE`?
These checks are for the thing status not the status given by unifi controller about the device.

scurb replied:
Exactly - I've fixed this and it will be commited during the day.

scurb replied:
Changed now.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -111,15 +114,34 @@ public void initialize() {
-            if (command == REFRESH) {
+        logger.debug("Handling command = {} for channel = {}", command, channelUID);
+
+        if (command instanceof OnOffType) {
```
Put this block in a separate method, just like is done with the `refreshChannel` call.

scurb replied:
Done.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -195,80 +218,85 @@ private boolean isClientHome(@Nullable UniFiClient client) {
         // mgb: only refresh if we're ONLINE
-        if (getThing().getStatus() == ONLINE) {
-            logger.debug("Refreshing channel = {}", channelUID);
+        // if (getThing().getStatus() == ONLINE) {
```
Why is this commented out?

scurb replied:
Changed - Thing status is now used again.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -230,12 +235,28 @@ private void cachePut(Map<String, UniFiClient> cache, UniFiClient client) {
+        return client;
+    }
+
+    public UniFiClient @Nullable [] setClientBlock(UniFiClient client) throws UniFiException {
```
These methods should not be in the class. It's just used here as a pass through, so this class should not have that knowledge. Instead create a method here to return the controller. Then the client handler can get the controller and directly call those methods.

scurb replied:
Moved!

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiClient.java:
```diff
@@ -50,6 +50,16 @@
 
+    protected boolean blocked;
+
+    public boolean getBlocked() {
```
These methods should be put in the right order not just in between fields.

Hilbrand replied:
Also rename:
```suggestion
    public boolean isBlocked() {
```

scurb replied:
Renamed.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -105,6 +107,54 @@ public void logout() throws UniFiException {
         executeRequest(req);
     }
 
+    public UniFiClient[] blockStation(UniFiClient client) throws UniFiException {
```
Don't copy paste methods, just create 1 method with a boolean parameter `blocked`.

Hilbrand replied:
The methods returns here something. But it's not used anywhere. So maybe the return is not necessary. Just make it void.

scurb replied:
Changed to void and only one function. State controlled via function parameter (boolean)

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -105,6 +107,54 @@ public void logout() throws UniFiException {
+    public UniFiClient[] blockStation(UniFiClient client) throws UniFiException {
+        UniFiControllerRequest<UniFiClient[]> req = newRequest(UniFiClient[].class);
+
+        String siteName = "";
```
sitename is a property of the handler therefor you should not try to find it here, but pass the sitename as parameter with this method. If due to some technical reason you really need to search for the sitename, it should be handled by the caller of this method.

scurb replied:
This is changed. Now given by a function.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -90,7 +93,7 @@ public void initialize() {
         }
 
-        UniFiControllerThingHandler controllerHandler = (UniFiControllerThingHandler) bridge.getHandler();
+        controllerHandler = (UniFiControllerThingHandler) bridge.getHandler();
```
If you make `controllerHandler` a field also take care of the other places in this class that use the value of field such that the code become better. Let other places also make use of this field.

scurb replied:
Think I found the rest - and updated.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -230,12 +235,28 @@ private void cachePut(Map<String, UniFiClient> cache, UniFiClient client) {
         }
 
         // mgb: short circuit
-        if (client == null || BooleanUtils.isNotTrue(client.isWireless()) || !belongsToSite(client, site)) {
```
Removal of the `isWireless` check seem harmless. Just wondering why it was here in the first place. Maybe it has something to do with general support for wired devices or a left over of refactoring. @mgbowman should be able to give exclusion about this.

mgbowman replied:
LGTM as I believe it was left over from the refactoring for "future support" of wired clients.

scurb commented:
Thanks for review comments! I have refactored things and updated accordingly. Hope I understood your comments correctly @Hilbrand

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -111,14 +114,30 @@ public void initialize() {
+
+        if (client != null && controllerHandler != null && controllerHandler.getController() != null) {
+            try {
+                String siteName = controllerHandler.getSiteNameForClient(client);
```
Not sure if you need to search for it. The sitename is also available in `config.getSite()`. This is what is used to in `getClient`. I would assume that should also work here.

scurb replied:
I think the config.getSite() is from configuration made in OH (which is optional). I wanted to get site name (not ID) from "running" state.

scurb replied:
I haven't made any changes here yet. Do you think the extra API calls for every refresh would be OK? Cause I think we only get "site ID" on the "/stat/sta"-call, which will result in additional calls to get the Site for this ID (which then will have Site Name included).

scurb commented:
I guess the next step here is for @mgbowman to decide if this is a welcomed contribution or not? 

Or do you see any additional changes in the PR @Hilbrand ?

mgbowman commented:
@martinvw @Hilbrand - I just got back in town late last night and will do my best to look over the code and test the changes on my local UniFi install later tonight.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -230,12 +237,17 @@ private void cachePut(Map<String, UniFiClient> cache, UniFiClient client) {
+    public String getSiteNameForClient(UniFiClient client) {
+        UniFiSite site = sitesCache.get(client.getSiteId());
+        return site.getName();
     }
```
See my comments above about adding `client.getSite().getName()`

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiClient.java:
```diff
@@ -50,6 +50,12 @@
+        return this.blocked;
+    }
+
     protected UniFiDevice device;
```
The getters / setters should not be mixed with the property declarations.

Should be something like...

```
public class UniFiClient {

    // properties

    protected boolean blocked;

    // getters / setters

    public boolean isBlocked() { 
        return blocked;
    }

    public void setBlocked(boolean blocked) {
        this.blocked = blocked;
    }

    ...

    // toString() overrides are usually found at the end of a class declaration
    public String toString() {
        ...
    }

}
```

scurb replied:
Changed!

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiClient.java:
```diff
@@ -102,8 +108,13 @@ public final Boolean isWireless() {
+    }
+
+    public void setBlocked(boolean value) {
+        this.blocked = value;
```
See my comment above - `toString()` is usually found at the end of a class declaration.

scurb replied:
Changed!

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -105,6 +106,20 @@ public void logout() throws UniFiException {
     }
 
+    public void blockStation(UniFiClient client, boolean block, String siteName) throws UniFiException {
+        UniFiControllerRequest<UniFiClient[]> req = newRequest(UniFiClient[].class);
```
Since we're not interested in the controller response, you can safely make this a `newRequest(Void.class)` request which will skip the JSON unmarshalling.

scurb replied:
Yes, the function now is a "void function".

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -174,6 +174,10 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+    public @Nullable UniFiController getController() {
+        return controller;
+    }
+
```
This is starting to get messy with the inter-dependencies of `UniFiClientThingHandler` depending on both `UniFiControllerThingHandler` and `UniFiController`.

I think I have a better way and sent @scurb a PM on the community to discuss potential collaboration on this PR. It will also help in the future when we want to add more "write" functions to the binding.

scurb commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiClientThingConfig.java:
```diff
@@ -53,7 +53,8 @@ public boolean isValid() {
-        return String.format("UniFiClientConfig{cid: '%s', site: '%s', considerHome: %d}", cid, site, considerHome);
+        return String.format("UniFiClientConfig{cid: '%s', site: '%s', considerHome: %d, blocked: %s}", cid, site,
+                considerHome);
     }
```
Ah, a leftover after a wrong placement. Removed now.

scurb commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -192,6 +196,9 @@ private void cachePut(Map<String, UniFiClient> cache, UniFiClient client) {
                         break;
+                    case CACHE_KEY_PREFIX_BLOCKED:
+                        suffix = Boolean.toString(client.isBlocked());
+                        break;
```
Ok, removed the block-parts from the lookup table!

scurb commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -85,8 +83,10 @@
     private static final String CACHE_KEY_PREFIX_ALIAS = "alias";
 
+    private static final String CACHE_KEY_PREFIX_BLOCKED = "blocked";
+
```
Ok, removed the block-parts from the lookup table - same as the other comment regarding this!

scurb commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -85,8 +83,10 @@
+
     private static final List<String> CACHE_KEY_PREFIXES = Arrays.asList(CACHE_KEY_PREFIX_MAC, CACHE_KEY_PREFIX_IP,
-            CACHE_KEY_PREFIX_HOSTNAME, CACHE_KEY_PREFIX_ALIAS);
+            CACHE_KEY_PREFIX_HOSTNAME, CACHE_KEY_PREFIX_ALIAS, CACHE_KEY_PREFIX_BLOCKED);
```
Ok, removed the block-parts from the lookup table - same as the other comment regarding this!

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -105,6 +106,20 @@ public void logout() throws UniFiException {
     }
 
+    public void blockStation(UniFiClient client, boolean block, String siteName) throws UniFiException {
+        UniFiControllerRequest<UniFiClient[]> req = newRequest(UniFiClient[].class);
```
I think the following was the idea: Since the return value is not used, it's not needed to parse the returned json either: 
```suggestion
        UniFiControllerRequest<Void> req = newRequest(Void.class);
```

scurb replied:
But of course! Updated.

mgbowman commented:
Just an update, I've reached out to @scurb and we're going to collaborate on this PR as there's a refactoring opportunity that will make this PR cleaner as well as future PRs for other features being requested in the community thread.

Please hold off on merging this PR as I want to put a call out to test the changes as there's quite a bit of code that is getting moved around for better class cohesion.

Hilbrand commented:
@mgbowman Is it the intention to add the refactoring to this pr or create a new pr?

mgbowman commented:
@Hilbrand I asked @scurb to give me commit access to his repo (he did) so it’s my intention to add it to this PR. 

If I understand github correctly, I should be able to push to his feature branch and this PR should update accordingly right?

Hilbrand commented:
@mgbowman Yes if you push to the feature branch this PR will be updated.

mgbowman commented:
I just pushed the refactoring changes. I redid the commit history to try and make the commits as small as possible. Unfortunately, the [Handler and Controller refactoring](https://github.com/openhab/openhab2-addons/pull/4845/commits/c353dd6dffc1a7771c79a6c0e95096aeb783f6ec) commit is the most complicated but was necessary 😏 

The refactoring has made implementing new features simple and more logical. I've already implemented `site` and `wiredClient` things but will wait to open another PR after this gets merged.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiController.java:
```diff
@@ -0,0 +1,300 @@
+        UniFiControllerRequest<UniFiSite[]> req = newRequest(UniFiSite[].class);
+        req.setPath("/api/self/sites");
+        UniFiSite[] sites = executeRequest(req);
+        logger.debug("Found {} UniFi Site(s): {}", sites.length, lazyFormatAsList(sites));
```
It looks like `executeRequest` can return null, which would lead to a null pointer exception here. Same on other places.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiController.java:
```diff
@@ -0,0 +1,300 @@
+        try {
+            result = request.execute();
+        } catch (UniFiExpiredSessionException e) {
+            login();
```
`login` calls `executeRequest` which means just looking at the code it can result in an endless loop. I don't know if it is practically possible, but it might be good to guard against an endless loop.

mgbowman replied:
The expired session exception should _theoretically_ only ever get thrown if you've already successfully logged in but your refresh value is so high that the session times out (will almost never happen).

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiController.java:
```diff
@@ -0,0 +1,300 @@
+            login();
+            result = executeRequest(request);
+        } catch (UniFiNotAuthorizedException e) {
+            logger.warn("Not Authorized! Please make sure your controller credentials have administrator rights");
```
If this happens, is that something the user must know? In other words, should this mean the thing should be set to offline with this message?

mgbowman replied:
Depends on your point of view. This error is simply indicating you don't have "admin" privileges. This will only ever happen if you try and "write" to a channel (`blocked`, `reconnect`) that requires admin.

martinvw replied:
It would be really nice if you could somehow feed this back to the thing-status, you don't have to mark it OFFLINE, but even when its ONLINE you can add a status line. It would be great to see in the paperUI why blocking fails

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/util/UniFiSiteInstanceCreator.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @author Matthew Bowman - Initial contribution
+ */
+
```
empty line

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiBaseThingHandler.java:
```diff
@@ -0,0 +1,133 @@
+                        try {
+                            handleCommand(entity, channelUID, command);
+                        } catch (UniFiException e) {
+                            logger.error("Unexpected error handling command = {} for channel = {} : {}", command,
```
Don't log to error, it's reserved within openHAB for system wide errors. Make it warn or debug.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/model/UniFiSite.java:
```diff
@@ -23,6 +23,12 @@
+    private final transient UniFiController controller;
+
+    public UniFiSite(UniFiController controller) {
+        this.controller = controller;
```
The `controller` field doesn't seem to be used here. Is it something for part of future support? Same question for other classes

mgbowman replied:
> Is it something for part of future support?

Exactly! The model classes have access to the controller directly which results in a cleaner API:

```java
public class UniFiSiteThingHandler {
  public void someMethod() {
    UniFiSite site = ...
    site.getTotalClients();
  }
}

public class UniFiSite {
  public long getTotalClients() {
    return controller.getTotalClients(this);
  }
}

public class UniFiController {
  public long getTotalClients(UniFiSite site) {
    ...
  }
}
```

This means all the `UniFiXXXThingHandlers` do not need direct access to the controller, just the entitles they are "managing". The controller access is encapsulated in the model classes.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/cache/UniFiCache.java:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @author Matthew Bowman - Initial contribution
+ */
+public abstract class UniFiCache<T> extends HashMap<String, T> {
```
I most cases it's better to implement this class that contains a HashMap instead of extending it. By extending it you also give access to all other methods and you really must be sure that all use cases actually handle the everything in HashMap correctly. Some would even call this an anti-pattern. If I did see it correctly you only use the methods you have overridden. Adding very little advantage to extend it anyway.

mgbowman replied:
I thought about this when I implemented it and I agree with you. I'll refactor this to delegate to an internal instance of `HashMap` instead of extending it.

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/cache/UniFiCache.java:
```diff
@@ -0,0 +1,94 @@
+
+    public static final String PREFIX_NAME = "name";
+
+    private final Logger logger = LoggerFactory.getLogger(UniFiCache.class);
```
Because this is an abstract class, you might want to create the logger with the subclass name to better identify it.
```suggestion
    private final Logger logger = LoggerFactory.getLogger(getClass());
```

mgbowman commented:
@Hilbrand I made all the requested changes and replied regarding the `UniFiExpiredSessionException` and `UniFiNotAuthorizedException`.

Is the automatic build failing because of the ongoing ESH reintegration?

Hilbrand commented:
Yes build fails due to ESH reintegration. The master branch this change is based on has a build issue due to moved packages. But the travis build succeeded so we know it can build.

mgbowman commented:
Should I rebase on the latest master and force push?

Hilbrand commented:
I would be nice. So we get all green (I hope, at least I think we should be able to see in the log of jenkins the build succeeded, something that fails currently with this pr)

mgbowman commented:
@martinvw can you give the code a review so we can get this merged?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/cache/UniFiCache.java:
```diff
@@ -0,0 +1,97 @@
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private Map<String, T> map = new HashMap<>();
```
Minor details: this field and the next one could be made final.
